examples/karpenter/MIGRATION.md

```diff
@@ -0,0 +1,249 @@
+
+1. Verify the Terraform changes:
+   ```bash
+   ./tf.sh all plan
+   ```
+
+2. Apply the Terraform changes:
+   ```bash
+   ./tf.sh all apply
+   ```
+
+3. Verify Karpenter installation:
+   ```bash
+   kubectl get pods -n karpenter
+   ```
+
+### 3. Configure Karpenter Node Classes and Node Pools
+
+1. Copy the `examples/karpenter` folder (this directory) onto your deployment directory.
+
+2. Update the node classes in `examples/karpenter/ec2nodeclasses/`:
+   1. You can run the following if you have `envsubst` installed:
+   ```bash
+    export EKS_NODES_ROLE_NAME=$(./tf.sh cluster output_json eks | jq -r '.eks.value.nodes.roles[0].name')
+    export EKS_CLUSTER_NAME=$(./tf.sh cluster output_json eks | jq -r '.eks.value.cluster.specs.name')
+    export AWS_REGION=$(./tf.sh cluster output_json eks | jq -r '.infra.value.region') ## This will be used later when scaling down the ASGs
+
+    for nc in karpenter/ec2nodeclasses/*yaml
+    do
+      cat $nc | envsubst > $nc.tmp && mv $nc.tmp $nc
+    done
+   ```
+   2. Or do it manually
+   - Update the role to match your EKS node role. Get value: `./tf.sh cluster output_json eks | jq '.eks.value.nodes.roles[0].name'`
+   - Update security group and subnet selectors to match your cluster. Should be the `<cluster_name>`. Get value: `./tf.sh cluster output_json eks | jq '.eks.value.cluster.specs.name'`
+   3. Verify the configuration for all the `ec2nodeclasses` and `nodepools`.
+
+3. :warning: Update the AMI selector in the nodepools. `al2023@latest` is not recommended for production environments. See karpenter docs [managing-amis](https://karpenter.sh/docs/tasks/managing-amis).
+
+4. Update the node pools in `examples/karpenter/nodepools/`:
+   - Match the instance types with your current node groups
+   - Configure capacity requirements.
+   - Most likely you will have `additional_node_groups`, you will need to create a <nodepool>.yaml for each. most likely it will just be an exact copy of compute/gpu with a different instance type(s)
+   - **[OPTIONAL]** You dont necessarily have to set the `instance_types`, see karpenter docs for additional information on how to configure [nodepools](https://karpenter.sh/docs/concepts/nodepools/).
+
+5. Apply the Karpenter configurations:
+   ```bash
+   bash -c karpenter/create-karpenter-configs.sh
+   ```
+
+### 5. Scale down Cluster autoscaler
+Scale down autoscaler so to avoid collision with karpenter. This is temporary while the nodes are being migrated.
+
+```bash
+kubectl scale deployments/cluster-autoscaler --replicas=0 -n domino-platform
+
+# verify that cas is not running
+kubectl get pods -n domino-platform | grep autoscaler
+```
+
+### 5. Verify Karpenter Setup
+
+1. Check Karpenter components:
+
+   ```bash
+   kubectl get pods -n karpenter
+   kubectl get ec2nodeclass
+   kubectl get nodepool
+   ```
+
+  1. Ensure the `ec2nodeclasses` and `nodepools` are `Ready`
+
+  ```bash
+    NAME                  READY   AGE
```

Comment:
I think this not being flush with the start of the backticks is making it display differently.